src/main/solution_space/path_finder.py

```diff
@@ -0,0 +1,166 @@
+                 f'Median of diff number is {median_of_diff_number_to_goal}')
+        graph_vertex = self.__find_closest_vertex_with_path(user_code, user, goal)
+        # We can have graph_vertex = None
+        if graph_vertex and self.__go_through_graph(user_code, graph_vertex, goal, median_of_diff_number_to_goal):
+            log.info(f'We are going through graph')
+            next_code = self.__apply_minimal_actions_number(user_code, graph_vertex.code)
+        else:
+            log.info(f'We are going directly to the goal')
+            next_code = self.__apply_minimal_actions_number(user_code, goal.code)
+        log.info(f'Finish finding the next code state. The user: {user}\nThe user code: {user_code}\n'
+                 f'The next code: {next_code}')
+        return next_code
+
+    # Sort candidates and return the best for user_code from ones
+    @staticmethod
+    def __choose_best_vertex(user_code: Code, user: User, vertices: List[Vertex]) -> Optional[Vertex]:
+        if len(vertices) == 0:
+            return None
+        candidates = list(map(lambda vertex: Candidate(user_code, vertex, user), vertices))
+        candidates.sort()
+        return candidates[-1].vertex
+
+    # Use '__choose_best_vertex' for choose the best goal from all ones
+    # Note: A goal is a vertex, which has the rate equals 1 and it connects to the end vertex
+    def __find_closest_goal(self, user_code: Code, user: User) -> Vertex:
+        return self.__choose_best_vertex(user_code, user, self._graph.end_vertex.parents)
+
+    # Calculate a set of vertices, which contains all vertices
+    # that have a distance to the goal <= than 'user_code' (number of diffs)
+    # Return None, if the size of the set is zero, otherwise run '__choose_best_vertex'
+    # Note: we have to remove the 'user_code' from the set
+    def __find_closest_vertex_with_path(self, user_code: Code, user: User, goal: Vertex) -> Optional[Vertex]:
+        user_diffs_to_goal = get_diffs_number(user_code.file_with_code, goal.code.file_with_code)
+        candidates = []
+        vertices = self._graph.get_traversal()
+        vertices.remove(self._graph.start_vertex)
+        for vertex in vertices:
+            # We don't want to add to result the same vertex
+            if are_asts_equal(user_code.ast, vertex.code.ast):
+                continue
+            diffs = get_diffs_number(vertex.code.file_with_code, goal.code.file_with_code)
+            if diffs <= user_diffs_to_goal:
+                candidates.append(vertex)
+        return self.__choose_best_vertex(user_code, user, vertices)
+
+    # Choose the best way to go to the goal
+    # For example, if we have a good way through the graph, we should advise it,
+    # but if don't we would advise going directly to the goal
+    @staticmethod
+    def __go_through_graph(user_code: Code, graph_vertex: Vertex, goal: Vertex,
+                           median_of_diff_number_to_goal: int) -> bool:
```

Comment:
maybe we should pass the graph as argument and get median_of_diff_number_to_goal inside the method? We don't use it anywhere else except there + our algorithm can change, but we probably need a graph in this method anyway